packages/core/src/external/sftp/surescripts/message.ts

```diff
@@ -0,0 +1,168 @@
+import { parseNameDemographics } from "./demographics";
+
+export function canGenerateSurescriptsMessage(
+  transmission: Transmission<TransmissionType>,
+  patients: Patient[]
+): boolean {
+  if (patients.length === 0) return false;
+  if (transmission.npiNumber == null || transmission.npiNumber.trim().length === 0) return false;
+  return true;
+}
+
+export function toSurescriptsPatientLoadFile(
+  client: SurescriptsSftpClient,
+  transmission: Transmission<TransmissionType>,
+  patients: Patient[]
+): Buffer {
+  const header = toSurescriptsPatientLoadRow(
+    {
+      recordType: "HDR",
+      version: SURESCRIPTS_VERSION,
+      usage: client.usage,
+      senderId: client.senderId,
+      senderPassword: client.senderPassword,
+      receiverId: client.receiverId,
+      patientPopulationId: transmission.cxId,
+      lookBackInMonths: 12,
+      transmissionId: transmission.id,
+      transmissionDate: transmission.date,
+      transmissionFileType: "PMA",
+      transmissionAction: "U",
+      fileSchedule: "ADHOC",
+    },
+    patientLoadHeaderSchema,
+    patientLoadHeaderOrder
+  );
+
+  const details = patients
+    .map((patient, index) => {
+      const { firstName, middleName, lastName, prefix, suffix } = parseNameDemographics(
+        patient.data
+      );
+      const gender = patient.data.genderAtBirth ?? "U";
+      const genderAtBirth = gender === "O" ? "U" : gender;
+
+      const address = Array.isArray(patient.data.address)
+        ? patient.data.address[0]
+        : patient.data.address;
+      if (!address) return null;
+
+      return toSurescriptsPatientLoadRow(
+        {
+          recordType: "PNM",
+          recordSequenceNumber: index + 1,
+          assigningAuthority: METRIPORT_OID,
+          patientId: patient.id,
+          lastName,
+          firstName,
+          middleName,
+          prefix,
+          suffix,
+          addressLine1: address.addressLine1,
+          addressLine2: address.addressLine2,
+          city: address.city,
+          state: address.state,
+          zip: address.zip,
+          dateOfBirth: patient.data.dob.replace(/-/g, ""), // TODO: check order
+          genderAtBirth,
+          npiNumber: transmission.npiNumber,
+        },
+        patientLoadDetailSchema,
+        patientLoadDetailOrder
+      );
+    })
+    .filter(Boolean) as Buffer[];
+
+  const footer = toSurescriptsPatientLoadRow(
+    {
+      recordType: "TRL",
+      totalRecords: details.length,
+    },
+    patientLoadFooterSchema,
+    patientLoadFooterOrder
+  );
+
+  return Buffer.concat([header, ...details, footer]);
+}
+
+export function toSurescriptsPatientLoadRow<T extends object>(
+  data: T,
+  schema: z.ZodObject<z.ZodRawShape>,
+  order: OutgoingFileRowSchema<T>
+): Buffer {
+  if (!schema.safeParse(data).success) {
+    console.log("Invalid data", data);
+  }
+  return Buffer.from(order.map(field => field.toSurescripts(data)).join("|"), "ascii");
```

Comment:
I suggest we avoid one liners w/ a lot of logic, it makes the code harder to read/reason about. IMO it's better to break into separated constants with the result of each operation. It also helps if you're debugging it, as you can hover over the constant and see each step of the operation.

Reply:
Awesome will definitely do this, totally agree that it is hard to read (even for me, who just wrote it yesterday!)